File: Management/Services/Canvas/CanvasServiceUtils.cs
using RestSharp;

namespace Management.Services.Canvas;

public class CanvasServiceUtils
{
  private const string BaseUrl = "https://snow.instructure.com/api/v1/";
  private readonly IWebRequestor webRequestor;

  public CanvasServiceUtils(IWebRequestor webRequestor)
  {
    this.webRequestor = webRequestor;
  }

  internal async Task<IEnumerable<T>> PaginatedRequest<T>(RestRequest request)
  {
    var requestCount = 1;
    request.AddQueryParameter("per_page", "100");
    var (data, response) = await webRequestor.GetAsync<T>(request);

    if (response.ErrorMessage?.Length > 0)
    {
      Console.WriteLine("error in response");
      Console.WriteLine(response.ErrorMessage);
      throw new Exception("error in response");
    }

    var returnData = data != null ? new T[] { data } : new T[] { };
    var nextUrl = getNextUrl(response.Headers);

    while (nextUrl is not null)
    {
      requestCount += 1;
      RestRequest nextRequest = new RestRequest(nextUrl);
      var (nextData, nextResponse) = await webRequestor.GetAsync<T>(nextRequest);
      if (nextData is not null)
        returnData = returnData.Append(nextData).ToArray();
      nextUrl = getNextUrl(nextResponse.Headers);
    }

    if (requestCount > 1)
      Console.WriteLine($"Requesting {typeof(T)} took {requestCount} requests");

    return returnData;
  }

  protected static string? getNextUrl(IEnumerable<HeaderParameter>? headers) =>
    headers
      ?.ToList()
      .Find(h => h.Name == "Link")
      ?.Value?.ToString()
      ?.Split(",")
      .Where(url => url.Contains("rel=\"next\""))
      .FirstOrDefault()
      ?.Split(";")
      .FirstOrDefault()
      ?.TrimEnd('>')
      .TrimStart('<')
      .Replace(" ", "")
      .Replace(BaseUrl, "");
}

File: nextjs/src/services/canvas/canvasServiceUtils.ts
// services/canvasServiceUtils.ts

import { AxiosResponseHeaders, RawAxiosResponseHeaders } from "axios";
import { axiosClient } from "../axiosUtils";

export const baseCanvasUrl = "https://snow.instructure.com";
export const canvasApi = baseCanvasUrl + "/api/v1";

const getNextUrl = (
  headers: AxiosResponseHeaders | RawAxiosResponseHeaders
): string | undefined => {
  const linkHeader: string | undefined =
    typeof headers.get === "function"
      ? (headers.get("link") as string)
      : ((headers as RawAxiosResponseHeaders)["link"] as string);

  if (!linkHeader) return undefined;

  const links = linkHeader.split(",").map((link) => link.trim());
  const nextLink = links.find((link) => link.includes('rel="next"'));

  if (!nextLink) return undefined;

  const nextUrl = nextLink.split(";")[0].trim().slice(1, -1);
  return nextUrl;
};

export const canvasServiceUtils = {
  async paginatedRequest<T>(request: { url: string }): Promise<T[]> {
    var requestCount = 1;
    const url = new URL(request.url);
    url.searchParams.set("per_page", "100");

    const { data: firstData, headers: firstHeaders } = await axiosClient.get<T>(
      url.toString()
    );

    var returnData: T[] = firstData ? [firstData] : [];
    var nextUrl = getNextUrl(firstHeaders);
    console.log("got first request", nextUrl, firstHeaders);

    while (nextUrl) {
      requestCount += 1;
      const { data, headers } = await axiosClient.get<T>(nextUrl);
      if (data) {
        returnData = [...returnData, data];
      }
      nextUrl = getNextUrl(headers);
    }

    if (requestCount > 1) {
      console.log(
        `Requesting ${typeof returnData} took ${requestCount} requests`
      );
    }

    return returnData;
  },
};
